src/aspire/basis/ffb_3d.py

```diff
@@ -115,32 +118,32 @@ def _precomp(self):
             ang_phi_wtd_odd.append(phi_wtd_m_odd)
 
         # evaluate basis function in the theta dimension
-        ang_theta = np.zeros((n_theta, 2 * self.ell_max + 1), dtype=theta.dtype)
+        ang_theta = xp.zeros((n_theta, 2 * self.ell_max + 1), dtype=theta.dtype)
 
-        ang_theta[:, 0 : self.ell_max] = np.sqrt(2) * np.sin(
-            theta @ m_reshape(np.arange(self.ell_max, 0, -1), (1, self.ell_max))
+        ang_theta[:, 0 : self.ell_max] = np.sqrt(2) * xp.sin(
+            theta @ m_reshape(xp.arange(self.ell_max, 0, -1), (1, self.ell_max))
         )
-        ang_theta[:, self.ell_max] = np.ones(n_theta, dtype=theta.dtype)
-        ang_theta[:, self.ell_max + 1 : 2 * self.ell_max + 1] = np.sqrt(2) * np.cos(
-            theta @ m_reshape(np.arange(1, self.ell_max + 1), (1, self.ell_max))
+        ang_theta[:, self.ell_max] = xp.ones(n_theta, dtype=theta.dtype)
+        ang_theta[:, self.ell_max + 1 : 2 * self.ell_max + 1] = np.sqrt(2) * xp.cos(
+            theta @ m_reshape(xp.arange(1, self.ell_max + 1), (1, self.ell_max))
         )
 
-        ang_theta_wtd = (2 * pi / n_theta) * ang_theta
+        ang_theta_wtd = (2 * np.pi / n_theta) * ang_theta
 
-        theta_grid, phi_grid, r_grid = np.meshgrid(
-            theta, phi, r, sparse=False, indexing="ij"
+        theta_grid, phi_grid, r_grid = xp.meshgrid(
+            theta.flatten(), phi.flatten(), r.flatten(), sparse=False, indexing="ij"
         )
-        fourier_x = m_flatten(r_grid * np.cos(theta_grid) * np.sin(phi_grid))
-        fourier_y = m_flatten(r_grid * np.sin(theta_grid) * np.sin(phi_grid))
-        fourier_z = m_flatten(r_grid * np.cos(phi_grid))
+        fourier_x = m_flatten(r_grid * xp.cos(theta_grid) * xp.sin(phi_grid))
+        fourier_y = m_flatten(r_grid * xp.sin(theta_grid) * xp.sin(phi_grid))
+        fourier_z = m_flatten(r_grid * xp.cos(phi_grid))
         fourier_pts = (
             2
-            * pi
-            * np.vstack(
+            * xp.pi
+            * xp.vstack(
                 (
-                    fourier_z[np.newaxis, ...],
-                    fourier_y[np.newaxis, ...],
-                    fourier_x[np.newaxis, ...],
+                    fourier_z[xp.newaxis, ...],
+                    fourier_y[xp.newaxis, ...],
+                    fourier_x[xp.newaxis, ...],
```

Comment:
`newaxis` is explicit and clear, but could change to `None` if we want to remove reference to the numerical package.

Reply:
tough call.  It would stink if numpy/cupy ever change `newaxis` to not be `None`.  I agree less reference to xp/np the better though. I'll change and push it up.

Reply:
yeah, thought about that. Can't imagine the amount of code that would break if that happened lol